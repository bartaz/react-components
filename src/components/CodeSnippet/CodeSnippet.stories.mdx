import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";
import { useState } from "react";

import CodeSnippet, { CodeSnippetBlock, CodeSnippetDropdown } from "./CodeSnippet";

<Meta
  title="CodeSnippet"
  component={CodeSnippet}
  argTypes={{
    className: {
      control: {
        type: "text",
      },
    },
  }}
/>

### Code snippet

This is a [React](https://reactjs.org/) component for the Vanilla [Code snippet](https://docs.vanillaframework.io/base/code#code-snippet).

### Props

<ArgsTable of={CodeSnippet} />

#### Blocks

Single `CodeSnippet` component can render multiple separate code blocks. Blocks are provided as an array via the `blocks` prop. Each block object defines values of props for each code block of the snippet:

<ArgsTable of={CodeSnippetBlock} />

### Default

<Canvas>
  <Story name="Default">
    <CodeSnippet blocks={[ {code: "Mode: all Settings: maas_url=http://192.168.122.1:5240/MAAS"} ]} />
  </Story>
</Canvas>

### Title

<Canvas>
  <Story name="Title">
    <CodeSnippet blocks={[ { title: "Output", code: "Mode: all Settings: maas_url=http://192.168.122.1:5240/MAAS"} ]} />
  </Story>
</Canvas>

### Multiple blocks

<Canvas>
  <Story name="Multiple blocks">
    <CodeSnippet blocks={[
      { title: "JavaScript", code: "console.log('Vanilla');"},
      { title: "Output", code: "Vanilla" },
    ]} />
  </Story>
</Canvas>


### Icon / Prompt

<Canvas>
  <Story name="Icon">
    <CodeSnippet blocks={[
       { title: "Install on Linux", appearance: "linuxPrompt", code: "snap install toto"},
       { title: "Install on Windows", appearance: "windowsPrompt", code: "snap install toto"},
       { title: "Get from the Store", appearance: "url", code: "http://snapcraft.io/toto"} ,
     ]}/>
  </Story>
</Canvas>

### Numbered

<Canvas>
  <Story name="Numbered">
    <CodeSnippet blocks={[
       { appearance: "numbered",
         code: `#!/bin/bash
set -eu . $CONJURE_UP_SPELLSDIR/sdk/common.sh
if [[ "$JUJU_PROVIDERTYPE" == "lxd" ]]; then
  debug "Running pre-deploy for $CONJURE_UP_SPELL"
  sed "s/##MODEL##/$JUJU_MODEL/" $(scriptPath)/lxd-profile.yaml | lxc profile edit "juju-$JUJU_MODEL" || exposeResult "Failed to set profile" $? "false"
fi
exposeResult "Successful pre-deploy." 0 "true"`
       },
     ]}/>
  </Story>
</Canvas>

### Wrap lines

<Canvas>
  <Story name="Wrap lines">
    <CodeSnippet blocks={[
       { numbered: true,
         wrapLines: true,
         code: `#!/bin/bash
set -eu . $CONJURE_UP_SPELLSDIR/sdk/common.sh
if [[ "$JUJU_PROVIDERTYPE" == "lxd" ]]; then
  debug "Running pre-deploy for $CONJURE_UP_SPELL"
  sed "s/##MODEL##/$JUJU_MODEL/" $(scriptPath)/lxd-profile.yaml | lxc profile edit "juju-$JUJU_MODEL" || exposeResult "Failed to set profile" $? "false"
fi
exposeResult "Successful pre-deploy." 0 "true"`
       },
     ]}/>
  </Story>
</Canvas>

### Dropdowns

Dropdown select menus can be added to the headers of code blocks to allow users to choose one of the options. The dropdowns options are passed via `dropdowns` property in the block options object.

Dropdown configuration object is structured as follows:
<ArgsTable of={CodeSnippetDropdown} />

With the `options` being an array of option properties compatible with the `Select` options (`[{ label: string, value: string | number }]`).

<Canvas>
  <Story name="Dropdowns">
  {() => {
    const [lang, setLang] = useState("html");
    const code = {
      js: `console.log("Example 1");`,
      css: `.p-heading--two { color: red; }`,
      html: `<h1 class="p-heading--two">How to use code snippets</h1>`,
    };
    return <CodeSnippet blocks={[
       {
         code: code[lang],
         dropdowns: [
           { options: [
              { value: "js", label: "JS" },
              { value: "css", label: "CSS" },
              { value: "html", label: "HTML", selected: true },
            ],
            onChange: (event) => { setLang(event.target.value) }
           }
         ]
       },
     ]} />;
  }}
  </Story>
</Canvas>

