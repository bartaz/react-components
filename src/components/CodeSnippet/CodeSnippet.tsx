import classNames from "classnames";
import PropTypes from "prop-types";
import React from "react";
import type { ReactNode } from "react";

export type DropdownOptionPros = {
  label: string;
  value: string | number;
};

export type CodeSnippetDropdownProps = {
  onChange?: (evt: React.SyntheticEvent) => void;
  options: Array<DropdownOptionPros>;
};

export const CodeSnippetDropdown = ({
  options,
  onChange,
}: CodeSnippetDropdownProps): JSX.Element => {
  return (
    <select className="p-code-snippet__dropdown" onChange={onChange}>
      {options.map(({ label, value, ...props }) => (
        <option value={value} key={value || label} {...props}>
          {label}
        </option>
      ))}
    </select>
  );
};

CodeSnippetDropdown.props = {
  onChange: PropTypes.func,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    })
  ).isRequired,
};

export const CodeSnippetBlock = ({
  code,
  title,
  numbered = false,
  icon = null,
  isWrapped = false,
  dropdowns,
}: CodeSnippetBlockProps): JSX.Element => {
  let className = "p-code-snippet__block";
  let numberedCode;

  if (numbered) {
    className += "--numbered";

    // wrap code lines in spans (and preserve the whitespace)
    const lines = code.split(/\r?\n/);
    numberedCode = lines.map((line, i) => (
      <React.Fragment key={`p-code-snippet__line-${i}`}>
        <span className="p-code-snippet__line">{line}</span>
        {"\n"}
      </React.Fragment>
    ));
  } else if (icon) {
    className += "--icon";
  }

  className = classNames(className, {
    "is-windows-prompt": icon === "windowsPrompt",
    "is-url": icon === "url",
    "is-wrapped": isWrapped,
  });

  const hasDropdowns = dropdowns && dropdowns.length;

  return (
    <>
      {(title || hasDropdowns) && (
        <div className="p-code-snippet__header">
          <h5 className="p-code-snippet__title">{title}</h5>
          {hasDropdowns && (
            <div className="p-code-snippet__dropdowns">
              {dropdowns.map((dropdown, i) => (
                <CodeSnippetDropdown
                  {...dropdown}
                  key={`code-snippet-dropdown-${i}`}
                />
              ))}
            </div>
          )}
        </div>
      )}
      <pre className={className}>
        <code>{numbered ? numberedCode : code}</code>
      </pre>
    </>
  );
};

CodeSnippetBlock.props = {
  code: PropTypes.string.isRequired,
  title: PropTypes.string,
  numbered: PropTypes.bool,
  icon: PropTypes.oneOf(["linuxPrompt", "windowsPrompt", "url"]),
  isWrapped: PropTypes.bool,
  dropdowns: PropTypes.array,
};

export type Props = {
  className?: string;
  children?: ReactNode;
  blocks: Array<CodeSnippetBlockProps>;
};

type CodeSnippetBlockProps = {
  code: string;
  title?: string;
  numbered?: boolean;
  icon?: "linuxPrompt" | "windowsPrompt" | "url";
  isWrapped?: boolean;
  dropdowns?: Array<CodeSnippetDropdownProps>;
};

const CodeSnippet = ({ className, blocks }: Props): JSX.Element => {
  return (
    <div className={classNames("p-code-snippet", className)}>
      {blocks.map((props, i) => (
        <CodeSnippetBlock
          key={`code-snippet-block-${i}`}
          {...props}
        ></CodeSnippetBlock>
      ))}
    </div>
  );
};

CodeSnippet.propTypes = {
  className: PropTypes.string,
  blocks: PropTypes.array,
};

export default CodeSnippet;
