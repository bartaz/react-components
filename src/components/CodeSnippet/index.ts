export { default } from "./CodeSnippet";
export { CodeSnippetBlockAppearance } from "./CodeSnippetBlock";
