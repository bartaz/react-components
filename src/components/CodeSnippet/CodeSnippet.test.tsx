import { shallow } from "enzyme";
import React from "react";

import CodeSnippet from "./CodeSnippet";

describe("CodeSnippet ", () => {
  it("can render", () => {
    const wrapper = shallow(<CodeSnippet blocks={[{ code: "Test" }]} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("can render snippet with a title", () => {
    const wrapper = shallow(
      <CodeSnippet blocks={[{ title: "Title", code: "Test" }]} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("can render snippet with multiple blocks", () => {
    const wrapper = shallow(
      <CodeSnippet
        blocks={[
          { title: "Title", code: "Test" },
          { title: "Title 2", code: "Test 2" },
        ]}
      />
    );
    expect(wrapper.find("CodeSnippetBlock").length).toBe(2);
  });
});
