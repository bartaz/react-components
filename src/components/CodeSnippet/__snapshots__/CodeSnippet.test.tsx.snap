// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeSnippet  can render 1`] = `
<div
  className="p-code-snippet"
>
  <CodeSnippetBlock
    code="Test"
    key="code-snippet-block-0"
  />
</div>
`;

exports[`CodeSnippet  can render snippet with title 1`] = `
<div
  className="p-code-snippet"
>
  <CodeSnippetBlock
    code="Test"
    key="code-snippet-block-0"
    title="Title"
  />
</div>
`;
